### PR TITLE
Add an experimentation service for Visual Studio

### LINKED ISSUE #11029: Define a NuGet experimentation service in Visual Studio
This will be convenient for https://github.com/NuGet/Home/issues/10678.

### PR DESCRIPTION
**Pull request by nkolev92**
## Bug

Fixes: https://github.com/NuGet/Home/issues/11029

Regression? Last working version:

## Description

There are a few experiments happening in NuGet, 2 to be exact, and a 3rd to be added in https://github.com/NuGet/Home/issues/10678. 

There's no a lot of common infrastructure right now, the work is just repeated a few times. 

This change replaces the existing usage with this new API.

cc @jebriede 

This technically does a minor part of https://github.com/NuGet/Home/blob/dev/proposed/2021/TransitiveDependenciesInPMUI.md#non-goals, `Add significant infrastructure for A/B experimentation
`
## PR Checklist

- [x] PR has a meaningful title
- [x] PR has a linked issue.
- [x] Described changes

- **Tests**
  - [x] Automated tests added
  - **OR**
  
  - [ ] Test exception
  - **OR**
  - [ ] N/A 

- **Documentation**
  
  - [ ] Documentation PR or issue filled
  - **OR**
  - [x] N/A

### CODE REVIEW

File: src/NuGet.Clients/NuGet.VisualStudio.Common/Experimentation/ExperimentationConstants.cs [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+using System;
+
+namespace NuGet.VisualStudio
+{
+    public class ExperimentationConstants
+    {
+        internal ExperimentationConstants(string flightFlag, string flightEnvironmentVariable)
+        {
+            FlightFlag = flightFlag ?? throw new ArgumentNullException(nameof(flightFlag));
+            FlightEnvironmentVariable = flightEnvironmentVariable ?? throw new ArgumentNullException(nameof(flightEnvironmentVariable));
```
Comment (zivkan) [inline src/NuGet.Clients/NuGet.VisualStudio.Common/Experimentation/ExperimentationConstants.cs]: In a discussion with @jebriede a few weeks ago, telemetry doesn't know when a customer overrides the flighting flag, hence the dashboard that allows us to compare baseline and experimental groups gets polluted with bad data.

So, allowing an experiment to not have an env var should not only be allowed, but should probably be preferred over having an env var that will pollute telemetry with the wrong data.

**Review by zivkan — APPROVED** [review]
> Note that I did not define an interface, as at this point I think it's over-engineering.
> The service also needs to be accessed statically, so I decided not introduce additional overhead through MEF.

It's no more over-engineering than `IEnvironmentVariableReader` is. You even used `TestEnvironmentVariableReader` in your tests.

Any code using this service, as is created in this PR, is not unit testable. Well, it can be, only by setting the process environment variable, which is global state. Therefore a lot of extra care has to be taken when writing tests that want to test the feature on/off, as it will break tests that run in parallel that expect the other global state.

So, just as `IEnvironmentVariableReader` makes code more easily testable, I think there should be an interface for `NuGetExperimentationService`, and there should not be a static instance, it should be constructor/property injectable, so that tests can be written more easily.

**Comment by nkolev92** [comment]
> Any code using this service, as is created in this PR, is not unit testable. Well, it can be, only by setting the process environment variable, which is global state. Therefore a lot of extra care has to be taken when writing tests that want to test the feature on/off, as it will break tests that run in parallel that expect the other global state.

Brushes itself is a static class and I wanted to avoid refactoring it for now. 

I totally see your point though, but I also think it'd be trivial to refactor it when needed. Currently the 2 codepaths that invoke the experimentation service are not unit testable at all.

File: src/NuGet.Clients/NuGet.VisualStudio.Common/Experimentation/ExperimentationConstants.cs (hunk 1 above)
Comment (nkolev92) [reply src/NuGet.Clients/NuGet.VisualStudio.Common/Experimentation/ExperimentationConstants.cs]: Sure, I can change that. It doesn't change any of the current data, but an interesting learning for future experiments :)

**Comment by nkolev92** [comment]
@zivkan 

Made it a MEF service.

**Review by erdembayar — APPROVED** [review]

**Pull request merged**